src/openvino.cc

```diff
@@ -1032,10 +1035,14 @@ ModelInstanceState::SetInputTensors(
 
     const int64_t batchn_byte_size = GetByteSize(input_datatype, batchn_shape);
 
-    bool collect_input =
-        ((input_buffer_count > 1) || (request_count > 1) ||
-         (batch_pad_size_ != 0));
-    if (collect_input) {
+    const char* input_buffer;
+    size_t buffer_byte_size;
+    TRITONSERVER_MemoryType memory_type;
+    int64_t memory_type_id;
+    bool contiguous_input = collector.GetInputBufferIfContiguous(
+        input_name, &input_buffer, &buffer_byte_size, &memory_type,
+        &memory_type_id);
+    if (!contiguous_input) {
```

Comment:
What is `batch_pad_size_`?

Reply:
See here: https://github.com/triton-inference-server/openvino_backend/blob/ae5e85b18406f9fa58603719d6805ee9cf72c302/src/openvino.cc#L805

So, once loaded we can not modify the IO dimensions of the network.. If max_batch_size is N then we load the network with the shape supporting full N batches. Later say if the request comes with batch size M (M < N), we pad the data with 0 bytes at the end essentially batching with null requests to match the max_batch_size..

Reply:
It is possible to integrate this into `ProceesInput()` of `BackendInputCollector`, and I think it is a good suggestion as then the backend doesn't need to handle the buffer copy explicitly when the contiguous buffer has different memory type. That being said, changes are still needed for other backends as some backends may need the input to be in specified buffer location (i.e. TF / TRT) while some may not. We can add meaning to `buffer` passed to `ProcessTensor()` that a `nullptr` should be interpreted as that the backend accepts any pointer returned by `BackendInputCollector`. `buffer_byte_size` will be used as input / output according to value of `buffer`.

So the updated signature will look like: 
```
void ProcessTensor(
      const char* input_name, char** buffer, size_t* buffer_byte_size,
      const TRITONSERVER_MemoryType memory_type, const int64_t memory_type_id);
```
And then what it does is:
1. if `buffer` is not `nullptr`, copy input to provided buffer
2. otherwise
a. If contiguous and no padding to max batch size and the memory type and type id matches, then return the pointer and size of the buffer.
b. otherwise, then create a new buffer and collect the data padding the data with specified bytes. Finally return the pointer to this new buffer and its size.

One downside is that `CPU_PINNED` and `CPU` are different types in Triton, while most of the backends don't care, so a copy from one to the other type will be perform even though it is not necessary. The previous way gives the backend more control over what they can do based on the input property returned. We can change the signature a bit more to make `memory_type` / `memory_type_id` to be input / output as well, but that is more convoluted because the behavior is not consistent: they will only be overwritten if `buffer == nullptr` and `memory_type == CPU or CPU_PINNED`

Reply:
Yes.. That's what I had in mind.. but was not aware of these complications. 

> We can change the signature a bit more to make memory_type / memory_type_id to be input / output as well, but that is more convoluted because the behavior is not consistent: they will only be overwritten

Can you elaborate the complications? What about an additional parameter to indicate strict memory_type which will prevent overwrites?

Reply:
Actually, we can't really change `ProcessTensor()` in a backward compatible way because the signature will need to be changed to let it return a buffer managed by input collector and the buffer's properties. (We can append the new parameters as optional but that seems strange). And the backends will have to reflect the change to use the new functionality anyway.

So maybe we should overload `ProcessTensor()` to be a better version of the utility, where the input parameters will be:

- the input name to be processed
- the backend provided destination buffer and its size (can be nullptr if final buffer should be managed by the collector)
- a list of memory type id pairs that the backend can handle (if backend has provided a buffer, then just the type and id of the buffer)

And the output parameters will be:

- the destination buffer and its size
- memory type and id of the buffer

Underneath the function:

- if input collection is not needed (no backend provided buffer, contiguous, no padding, input type and id matches one of allowed type / id pair), it will just set the output parameters based on the contiguous buffer
- otherwise, if backend doesn't provide a destination buffer, it creates one with the type and id from the first allowed pair, and then reuse the same code as current `ProcessTensor()`

Reply:
I was going to say that we need to preserve backwards compat in the backend utils. I have a question about the last point. If the collector creates a destination buffer how is the lifetime managed? Is it deleted when the collector is destroyed? Will that mean we need to increase the lifetime of the collector for some backends?

Reply:
I was thinking that the created buffer will have the same lifetime as the input collector, so we will need to make sure it is valid until the model execution is completed, which I think our backends already doing that (will need to double check).

Reply:
The solution with overloaded `ProcessTensor` looks good and I agree we must keep the lifetime of the buffer limited to the input collector object.